src/array/union.rs

```diff
@@ -333,6 +336,127 @@ where
     }
 }
 
+/// Types that return a constructed rust `enum` by advancing
+/// iterator(s) of variants of a union array given the
+/// `type_id` for the variant.
+pub trait TypeIdIterator {
+    /// The rust enum
+    type Enum;
+
+    /// Advances the variant iterator(s) to get the next
+    /// value for the `type_id`.
+    fn next(&mut self, type_id: i8) -> Option<Self::Enum>;
```

Comment:
I did consider that, but, I went with the `Option` of `Item` to follow std `Iterator`'s design because internally that's exactly what this is, and the caller of this function will be much more aware of whether the underlying iterator is supposed to contain stuff or not at a certain point, and thus the decision of `panic`-ing could be delegated to the caller.

My view of this trait is just a handle to the iterator of the underlying variant given the type_id.

To me, calling a function that advances a rust iterator to always produce valid values and panic if its empty felt a bit unidiomatic -- because even in this case, the caller (`struct UnionArrayIntoIter`) has appropriate information about arrow standard validity.

Reply:
> To me, calling a function that advances a rust iterator to always produce valid values and panic if its empty felt a bit unidiomatic 

I agree, which is why I'd like to think about these traits as (hidden) implementation details. The generated iterators are always supposed to be used in combination with the iterator over `type_id`s i.e. the caller is always going to be `UnionArrayIntoIter`.

But I don't have a super strong opinion about this, so let's just keep it like this.

Reply:
Haven't updated the trait definition in the changes yes -- lets think a bit more and do that as a follow up?